plugin/storage/grpc/shared/archive.go

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2020 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shared
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/jaegertracing/jaeger/model"
+	"github.com/jaegertracing/jaeger/proto-gen/storage_v1"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+)
+
+// ArchiveReader wraps storage_v1.ArchiveSpanReaderPluginClient into spanstore.Reader
+type ArchiveReader struct {
+	client storage_v1.ArchiveSpanReaderPluginClient
+}
+
+// ArchiveWriter wraps storage_v1.ArchiveSpanWriterPluginClient into spanstore.Writer
+type ArchiveWriter struct {
+	client storage_v1.ArchiveSpanWriterPluginClient
+}
+
+// GetTrace takes a traceID and returns a Trace associated with that traceID from Archive Storage
+func (r *ArchiveReader) GetTrace(ctx context.Context, traceID model.TraceID) (*model.Trace, error) {
+	stream, err := r.client.GetArchiveTrace(upgradeContextWithBearerToken(ctx), &storage_v1.GetTraceRequest{
+		TraceID: traceID,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("plugin error: %w", err)
```

Comment:
I think this may mangle the NotFound error

Reply:
I copy that code from https://github.com/jaegertracing/jaeger/blob/0a889e8ea7ca887f30eaa814456d86ad97be3f8e/plugin/storage/grpc/shared/grpc_client.go#L69-L75
We have to fix it in two places if it contain a bug.

Reply:
should probably extract grpc status from error and compare it with codes.NotFound in both places

Though, the test seems to be wrong in the original grpc client span-reader `GetTrace` since mock returns `spanstore.ErrTraceNotFound`, but real client will always return grpc status: https://github.com/jaegertracing/jaeger/blob/66f9ca761a26715e05a533fcfd253fb353c4b7e2/plugin/storage/grpc/shared/grpc_client_test.go#L204-L211